src/CollectTestExecutionMemoryFootprints.php

```diff
@@ -0,0 +1,132 @@
+    implements BeforeTestHook,
+    AfterSuccessfulTestHook,
+    AfterLastTestHook,
+    TestListener
+{
+    /** @var array<string, array<int, int>> */
+    private $preTestMemoryUsages = [];
+
+    /** @var array<string, array<int, int>> */
+    private $postTestMemoryUsages = [];
+
+    public function startTestSuite(TestSuite $suite) : void
+    {
+        $suite->addTest(new EmptyBaselineMemoryUsageTest(EmptyBaselineMemoryUsageTest::TEST_METHOD));
+    }
+
+    public function executeBeforeTest(string $test) : void
+    {
+        \gc_collect_cycles();
+
+        $this->preTestMemoryUsages[$test][] = \memory_get_usage();
+    }
+
+    public function executeAfterSuccessfulTest(string $test, float $time) : void
+    {
+        \gc_collect_cycles();
+
+        $this->postTestMemoryUsages[$test][] = \memory_get_usage();
+    }
+
+    public function executeAfterLastTest() : void
+    {
+        if (! (
+            array_key_exists(EmptyBaselineMemoryUsageTest::class . '::' . EmptyBaselineMemoryUsageTest::TEST_METHOD, $this->preTestMemoryUsages)
+            && array_key_exists(EmptyBaselineMemoryUsageTest::class . '::' . EmptyBaselineMemoryUsageTest::TEST_METHOD, $this->postTestMemoryUsages)
+        )) {
+            throw new \Exception('Could not find baseline test: impossible to determine PHPUnit base memory overhead');
+        }
+
+        $baselineMemoryUsage = MeasuredBaselineTestMemoryLeak::fromBaselineTestMemoryUsages(
+            $this->preTestMemoryUsages[EmptyBaselineMemoryUsageTest::class . '::' . EmptyBaselineMemoryUsageTest::TEST_METHOD],
+            $this->postTestMemoryUsages[EmptyBaselineMemoryUsageTest::class . '::' . EmptyBaselineMemoryUsageTest::TEST_METHOD]
+        );
+
+        unset(
+            $this->preTestMemoryUsages[EmptyBaselineMemoryUsageTest::class . '::' . EmptyBaselineMemoryUsageTest::TEST_METHOD],
+            $this->postTestMemoryUsages[EmptyBaselineMemoryUsageTest::class . '::' . EmptyBaselineMemoryUsageTest::TEST_METHOD]
+        );
+
+        $memoryUsages = array_combine(
+            array_keys($this->preTestMemoryUsages),
+            array_map(
+                [MeasuredTestRunMemoryLeak::class, 'fromTestMemoryUsages'],
+                $this->preTestMemoryUsages,
+                $this->postTestMemoryUsages
+            )
+        );
+
+        $leaks = array_filter(array_map(function (MeasuredTestRunMemoryLeak $profile) use ($baselineMemoryUsage) : bool {
+            return $profile->leaksMemory($baselineMemoryUsage);
+        }, $memoryUsages));
+
+        if ([] !== $leaks) {
+            throw new \Exception(sprintf(
+                "The following test produced memory leaks:\n * %s\n",
+                implode("\n * ", array_keys($leaks))
+            ));
+        }
+
+        var_dump($memoryUsages, $leaks, $baselineMemoryUsage);
+    }
+
+    // The following bits are just stubbing the implementation of the
+    // deprecated {@see \PHPUnit\Framework\TestListener\TestListener} API
```

Comment:
Also, why not extend the Abstract class?

Reply:
Less coupling.